File: session_cpp/src/color.cpp
#include "color.h"

namespace session_cpp {

///////////////////////////////////////////////////////////////////////////////////////////
// Operators
///////////////////////////////////////////////////////////////////////////////////////////

/// Convert point to string representation
std::string Color::to_string() const {
  return fmt::format("Color({}, {}, {}, {}, {})", r, g, b, a, name);
}

/// Equality operator
bool Color::operator==(const Color &other) const {
  return r == other.r && g == other.g && b == other.b && a == other.a &&
         name == other.name;
}

/// Inequality operator
bool Color::operator!=(const Color &other) const { return !(*this == other); }

///////////////////////////////////////////////////////////////////////////////////////////
// JSON
///////////////////////////////////////////////////////////////////////////////////////////

nlohmann::ordered_json Color::to_json_data() const {
  return nlohmann::ordered_json{{"type", "Color"},
                                {"guid", guid},
                                {"name", name},
                                {"r", static_cast<int>(r)},
                                {"g", static_cast<int>(g)},
                                {"b", static_cast<int>(b)},
                                {"a", static_cast<int>(a)}};
}

Color Color::from_json_data(const nlohmann::json &data) {
  Color color = Color(static_cast<unsigned int>(data["r"]),
                      static_cast<unsigned int>(data["g"]),
                      static_cast<unsigned int>(data["b"]),
                      static_cast<unsigned int>(data["a"]), data["name"]);
  color.guid = data["guid"];
  return color;
}

void Color::to_json(const std::string &filepath) const {
  std::ofstream file(filepath);
  file << to_json_data().dump(4);
}

///////////////////////////////////////////////////////////////////////////////////////////
// Details
///////////////////////////////////////////////////////////////////////////////////////////

Color Color::from_json(const std::string &filepath) {
  std::ifstream file(filepath);
  nlohmann::json data;
  file >> data;
  return from_json_data(data);
}

Color Color::white() { return Color(255, 255, 255, 255, "white"); }

Color Color::black() { return Color(0, 0, 0, 255, "black"); }

std::array<double, 4> Color::to_float_array() const {
  return {r / 255.0, g / 255.0, b / 255.0, a / 255.0};
}

Color Color::from_float(double r, double g, double b, double a) {
  return Color(static_cast<unsigned int>(r * 255.0 + 0.5),
               static_cast<unsigned int>(g * 255.0 + 0.5),
               static_cast<unsigned int>(b * 255.0 + 0.5),
               static_cast<unsigned int>(a * 255.0 + 0.5));
}

///////////////////////////////////////////////////////////////////////////////////////////
// Not class methods
///////////////////////////////////////////////////////////////////////////////////////////

std::ostream &operator<<(std::ostream &os, const Color &color) {
  return os << color.to_string();
}

} // namespace session_cpp

File: session_rust/src/color.rs
use std::fmt;
use serde::{Deserialize, Serialize, ser::Serialize as SerTrait};
use uuid::Uuid;

/// A color with RGBA values.
///
/// # Fields
/// * `r` - Red component (0-255)
/// * `g` - Green component (0-255)
/// * `b` - Blue component (0-255)
/// * `a` - Alpha component (0-255)
/// * `guid` - Unique identifier
/// * `name` - Color name
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename = "Color")]
pub struct Color {
    pub guid: Uuid,
    pub name: String,
    pub r: u8,
    pub g: u8,
    pub b: u8,
    pub a: u8,
}

impl Color {
    /// Create new color.
    pub fn new(r: u8, g: u8, b: u8, a: u8) -> Self {
        Color { 
            guid: Uuid::new_v4(),
            name: "Color".to_string(),
            r, 
            g, 
            b, 
            a 
        }
    }

    /// Create white color.
    pub fn white() -> Self {
        let mut color = Color::new(255, 255, 255, 255);
        color.name = "white".to_string();
        color
    }

    /// Create black color.
    pub fn black() -> Self {
        let mut color = Color::new(0, 0, 0, 255);
        color.name = "black".to_string();
        color
    }

    /// Convert to float array [0-1].
    pub fn to_float_array(&self) -> [f32; 4] {
        [self.r as f32 / 255.0, self.g as f32 / 255.0, self.b as f32 / 255.0, self.a as f32 / 255.0]
    }

    /// Create from float values [0-1].
    pub fn from_float(r: f32, g: f32, b: f32, a: f32) -> Self {
        Color::new(
            (r * 255.0).round() as u8,
            (g * 255.0).round() as u8,
            (b * 255.0).round() as u8,
            (a * 255.0).round() as u8,
        )
    }

    /// Serialize to JSON string (for cross-language compatibility)
    pub fn to_json_data(&self) -> Result<String, Box<dyn std::error::Error>> {
        let mut buf = Vec::new();
        let formatter = serde_json::ser::PrettyFormatter::with_indent(b"    ");
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, formatter);
        SerTrait::serialize(self, &mut ser)?;
        Ok(String::from_utf8(buf)?)
    }

    /// Deserialize from JSON string (for cross-language compatibility)
    pub fn from_json_data(json_data: &str) -> Result<Self, Box<dyn std::error::Error>> {
        Ok(serde_json::from_str(json_data)?)
    }

    /// Serialize to JSON file
    pub fn to_json(&self, filepath: &str) -> Result<(), Box<dyn std::error::Error>> {
        let json = self.to_json_data()?;
        std::fs::write(filepath, json)?;
        Ok(())
    }

    /// Deserialize from JSON file
    pub fn from_json(filepath: &str) -> Result<Self, Box<dyn std::error::Error>> {
        let json = std::fs::read_to_string(filepath)?;
        Self::from_json_data(&json)
    }
}

impl Default for Color {
    fn default() -> Self {
        Self::white()
    }
}

impl fmt::Display for Color {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Color(r={}, g={}, b={}, a={}, name={})", self.r, self.g, self.b, self.a, self.name)
    }
}

